tsfc/coffee_mode.py

```diff
@@ -105,38 +141,33 @@ def find_optimal_atomics(monomials, argument_indices):
 
     :returns: list of atomic GEM expressions
     """
-    atomic_index = defaultdict(partial(next, itertools.count()))  # atomic -> int
-    connections = []
-    # add connections (list of tuples, items in each tuple form a product)
+    atomics = OrderedDict()
+    idx = 0
     for monomial in monomials:
-        connections.append(tuple(map(lambda a: atomic_index[a], monomial.atomics)))
-
-    if len(atomic_index) <= 1:
-        return tuple(iterkeys(atomic_index))
-
-    # set up the ILP
-    import pulp as ilp
-    ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
-    ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
-
-    # Objective function
-    # Minimise number of factors to pull. If same number, favour factor with larger extent
-    penalty = 2 * max(index_extent(atomic, argument_indices) for atomic in atomic_index) * len(atomic_index)
-    ilp_prob += ilp.lpSum(ilp_var[index] * (penalty - index_extent(atomic, argument_indices))
-                          for atomic, index in iteritems(atomic_index))
-
-    # constraints
-    for connection in connections:
-        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
-
-    ilp_prob.solve()
-    if ilp_prob.status != 1:
-        raise RuntimeError("Something bad happened during ILP")
-
-    def optimal(atomic):
-        return ilp_var[atomic_index[atomic]].value() == 1
+        for atomic in monomial.atomics:
+            if atomic not in atomics:
+                atomics[atomic] = idx
+                idx += 1
```

Comment:
Why was it a good idea to spell out `defaultdict(partial(next, itertools.count()))`? You don't even really need this mapping anymore. Just take:
```python
atomics = tuple(OrderedDict.fromkeys(*[monomial.atomics for monomial in monomials]))
```